### PR TITLE
refactor: more strict checks

### PR DESCRIPTION
**Pull request by viceice**
## Changes:
- depends on #14154

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/config/defaults.ts [hunk 1]
```diff
@@ -1,14 +1,14 @@
 import { getOptions } from './options';
 import type { AllConfig, RenovateOptions } from './types';
 
-const defaultValues = {
+const defaultValues: Record<string, unknown> = {
   boolean: true,
   array: [],
   string: null,
   object: null,
 };
```
Comment (JamieMagee) [inline lib/config/defaults.ts:9]: What about `as const`?

File: lib/config/index.ts [hunk 2]
```diff
@@ -37,7 +43,7 @@ export function filterConfig(
   const stages = ['global', 'repository', 'package', 'branch', 'pr'];
```
Comment (JamieMagee) [inline lib/config/index.ts:43]: This isn't related to your changes, but I am seeing this pattern in a few places, and wanted to get your opinion.

We have the stages defined here:

https://github.com/renovatebot/renovate/blob/1f28cfd2e5000d6c9988f9e32790e3fb5dfc50b1/lib/config/types.ts#L6-L11

But they're a type union, and types aren't there at runtime. What about this sort of pattern

```ts
export const renovateConfigStages = ['global', 'repository', 'package', 'branch', 'pr'] as const
export type RenovateConfigStage = typeof renovateConfigStages[number];
```

It means we can iterate over the stages at runtime, and the types are always up-to-date. What do you think?

(I'm not suggesting you change this here)

File: lib/config/migrations/custom/compatibility-migration.ts [hunk 3]
```diff
@@ -6,8 +6,8 @@ export class CompatibilityMigration extends AbstractMigration {
   override readonly propertyName = 'compatibility';
 
   override run(value: unknown): void {
-    if (is.object(value)) {
-      this.setSafely('constraints', value);
+    if (is.plainObject(value)) {
+      this.setSafely('constraints', value as Record<string, string>);
```
Comment (JamieMagee) [inline lib/config/migrations/custom/compatibility-migration.ts:10]:
```suggestion
    if (is.plainObject<string>(value)) {
      this.setSafely('constraints', value);
```

File: lib/config/migrations/custom/trust-level-migration.ts [hunk 4]
```diff
@@ -1,6 +1,7 @@
 import { AbstractMigration } from '../base/abstract-migration';
+import type { DeprecatedRenovateConfig } from '../types';
 
-export class TrustLevelMigration extends AbstractMigration {
+export class TrustLevelMigration extends AbstractMigration<DeprecatedRenovateConfig> {
```
Comment (JamieMagee) [inline lib/config/migrations/custom/trust-level-migration.ts:4]: I don't think this change is necessary. You made `DeprecatedRenovateConfig` the default.

File: lib/config/options/index.spec.ts [hunk 5]
```diff
@@ -38,7 +38,7 @@ describe('config/options/index', () => {
       .filter((option) => option.supportedPlatforms)
       .forEach((option) => {
         expect(option.supportedPlatforms).toBeNonEmptyArray();
-        for (const item of option.supportedPlatforms) {
+        for (const item of option.supportedPlatforms as string[]) {
```
Comment (JamieMagee) [inline lib/config/options/index.spec.ts:41]: nit: wrap this in `is.nonEmptyArray` or `        assert.nonEmptyArray(option.supportedManagers)`

File: lib/config/presets/internal/index.spec.ts [hunk 6]
```diff
@@ -4,13 +4,14 @@ import { massageConfig } from '../../massage';
...
 import * as _npm from '../npm';
+import type { Preset } from '../types';
 import * as internal from '.';
 
 jest.mock('./npm');
 jest.mock('../../../datasource/npm');
 
 const npm = mocked(_npm);
-npm.getPreset = jest.fn((_) => null);
+npm.getPreset = jest.fn((_) => Promise.resolve<Preset>(null as never));
```
Comment (JamieMagee) [inline lib/config/presets/internal/index.spec.ts:14]: nit: `mockFn` from `jest-mock-extended` is a shortcut for this.

```suggestion
npm.getPreset = mockFn();
```

File: lib/config/presets/util.spec.ts [hunk 7]
```diff
@@ -1,61 +1,61 @@
 import type { FetchPresetConfig, Preset } from './types';
 import { PRESET_DEP_NOT_FOUND, PRESET_NOT_FOUND, fetchPreset } from './util';
 
+const fetch = jest.fn(() => Promise.resolve<Preset>({}));
```
Comment (JamieMagee) [inline lib/config/presets/util.spec.ts:4]: Same here `mockFn()`

**Comment by JamieMagee** [comment]
I got about halfway through. I'll come back and review more later 👍

**Comment by viceice** [comment]
ok, will split smaller parts out, as the `RenovateConfig` are pretty big and a more longtime refactor. 🙃

File: lib/config/options/index.spec.ts (hunk 5 above)
Comment (viceice) [reply lib/config/options/index.spec.ts:41]: can you make a suggestion please, didn't get it working nicely

File: lib/config/defaults.ts (hunk 1 above)
Comment (viceice) [reply lib/config/defaults.ts:9]: Can you make a suggestion?

File: lib/config/index.ts (hunk 2 above)
Comment (viceice) [reply lib/config/index.ts:43]: Sounds good, but `renovateConfigStages` isn't a type only import then?

File: lib/config/migrations/custom/trust-level-migration.ts (hunk 4 above)
Comment (viceice) [reply lib/config/migrations/custom/trust-level-migration.ts:4]: Need to think how to optimize futher. Not reallly happy with that yet.

**Pull request closed without merging**